### PR TITLE
INGM-254 Combine test reports

### PR DESCRIPTION
**Pull request by pzinemanas**
### Description

This PR adds a script to combine XML reports from several runs of the tests. This is relevant to combine information from tests with different protocols/slaves. The script concatenates error/failure/skip messages, including a header with the protocol and slave information. The combined test result is defined as follows:

PASSED → If the test passed for all protocols and all slaves.
SKIPPED → If the test was skipped for all protocols and all slaves.
ERROR → If there is at least one error.
FAILED → If there is at least one failure if no error occurred.

Fixes # INGM-254

### Type of change

- [x] Add the `tests/combine_reports.py script`
- [x] Add a fixture to save the protocol and slave for each test session. 
- [x] Update Jenkinsfile to combine the reports before publishing them using JUnit.

### Tests
- [x] The script was verified using Jenkins (branch test-INGM-254).

[INGM-254]: https://ingeniamc.atlassian.net/browse/INGM-254?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[INGM-254]: https://ingeniamc.atlassian.net/browse/INGM-254?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/combine_reports.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+                        result = "ERROR" # ERROR before
+                    elif report_dict["testcases"][classname][name]["result"] == "ERROR":
+                        result = "ERROR" # ERROR now
+                    elif combined_report["testcases"][classname][name]["result"] == "FAILED":
+                        result = "FAILED" # FAILED before
+                    elif report_dict["testcases"][classname][name]["result"] == "FAILED":
+                        result = "FAILED" # FAILED now
+                    elif combined_report["testcases"][classname][name]["result"] == "SKIPPED":
+                        # SKIPPED before
+                        if combined_report["testcases"][classname][name]["result"] == "PASSED":
```
Comment (GullinBustin) [inline tests/combine_reports.py:145]: I don't understand this if.
combined_report["testcases"][classname][name]["result"] can't be "SKIPPED" and "PASSED" at the same time

File: tests/combine_reports.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+            combined_report["timestamp"] = report_dict["timestamp"]
+            combined_report["hostname"] = report_dict["hostname"]
+        combined_report["time"] += report_dict["time"]
+        combined_report["tests"] = report_dict["tests"]
+
+        for classname in report_dict["testcases"].keys():
+            if classname not in combined_report["testcases"]:
+                combined_report["testcases"][classname] = {}
+            for name in report_dict["testcases"][classname].keys():
+                if name not in combined_report["testcases"][classname]:
```
Comment (GullinBustin) [inline tests/combine_reports.py:108]: It's hard to understand what means these if and else, can you create some functions to encapsulate some of this logic? the function name will help to understand what it does

Comment (pzinemanas) [reply tests/combine_reports.py:123]: Ok, I just created a function to encapsulate this logic.

File: tests/combine_reports.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                        "result": result,
+                        "message": message,
+                        "output": output,
+                        "time": combined_report["testcases"][classname][name]["time"] + report_dict["testcases"][classname][name]["time"]
+                    }
+
+    for classname in combined_report["testcases"].keys():
+        for name in combined_report["testcases"][classname].keys():
+            result = combined_report["testcases"][classname][name]["result"]
+            if result == "PASSED":
```
Comment (MartinALL91) [inline tests/combine_reports.py:184]: I think you can use a dictionary to make it cleaner:

```
            result_dict = {
                'PASSED': 'passed',
                'ERROR': 'errors',
                'FAILED': 'failures',
                'SKIPPED': 'skipped'
            }
            combined_report[result_dict[result]] += 1
```

**Review by MartinALL91 — CHANGES_REQUESTED** [review]

File: Jenkinsfile [hunk 4]
```diff
@@ -172,6 +167,11 @@ pipeline {
                             venv\\Scripts\\python.exe -m coverage xml --include=ingeniamotion/*
                         '''
                         publishCoverage adapters: [coberturaReportAdapter('coverage.xml')]
+                        unstash 'test_reports'
+                        bat '''
+                            venv\\Scripts\\python.exe -m tests.combine_reports -i pytest_ethernet_0_report.xml pytest_ethernet_1_report.xml pytest_canopen_0_report.xml pytest_canopen_1_report.xml pytest_ethercat_0_report.xml pytest_ethercat_1_report.xml -o combined_report.xml
```
[the remark this answers is not in the document]
Comment (pzinemanas) [reply Jenkinsfile]: Done!

File: tests/combine_reports.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+                skip.set("message", testcase_dict["message"]) 
+                skip.text = testcase_dict["output"] 
+
+    dom = minidom.parseString(ET.tostring(tree, encoding='utf-8'))
+    with open(output_file, "wb") as f:
+        f.write(dom.toprettyxml(indent='\t').encode())
+
+
+def main(args):
+    if len(args.input_reports) < 2:
```
[the remark this answers is not in the document]
Comment (pzinemanas) [reply tests/combine_reports.py]: Here I check that more than two input files are passed in the -i argument.

[the remark this answers is not in the document]
Comment (MartinALL91) [reply tests/combine_reports.py]: Ok, I see. I thought that you were checking that both args are being passed.

**Review by GullinBustin — APPROVED** [review]

**Review by WoodElv — APPROVED** [review]

**Review by MartinALL91 — APPROVED** [review]

**Pull request merged**